staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -87,43 +87,56 @@ func (a customResourceStrategy) PrepareForCreate(ctx context.Context, obj runtim
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
 func (a customResourceStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) || a.status == nil {
-		return
-	}
-
 	newCustomResourceObject := obj.(*unstructured.Unstructured)
 	oldCustomResourceObject := old.(*unstructured.Unstructured)
 
 	newCustomResource := newCustomResourceObject.UnstructuredContent()
 	oldCustomResource := oldCustomResourceObject.UnstructuredContent()
 
-	// update is not allowed to set status
-	_, ok1 := newCustomResource["status"]
-	_, ok2 := oldCustomResource["status"]
-	switch {
-	case ok2:
-		newCustomResource["status"] = oldCustomResource["status"]
-	case ok1:
-		delete(newCustomResource, "status")
+	var incrementGeneration bool
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) && a.status != nil {
+		// If the /status subresource endpoint is installed, it
+		// indicates that the user wants to participate the spec/status
+		// convention, then only changes to `.spec` increments the
+		// `.metadata.generation`
+		newSpec, ok1 := newCustomResource["spec"]
+		oldSpec, ok2 := oldCustomResource["spec"]
+
+		// spec is changed, created or deleted
+		if (ok1 && ok2 && !apiequality.Semantic.DeepEqual(oldSpec, newSpec)) || (ok1 && !ok2) || (!ok1 && ok2) {
+			incrementGeneration = true
+		}
+	} else {
+		// except for the changes to `metadata`, any other changes
+		// cause the generation to increment.
+		newObjCopy := newCustomResourceObject.DeepCopy()
```

Comment:
Are the scalability folks still policing new deep copies as much as they used to? I can think of a couple ways we might avoid them but it's not pretty. For example, you could do a shallow copy while omitting `metadata`, or make a helper that knows how to ignore `metadata` while checking for mismatches in other top-level fields on either side of the diff. I wouldn't think it's worthwhile, but I've been told in the past to avoid deep copy whenever possible.